### ISSUE TITLE
object-shorthand rule and arrow functions

### ISSUE THREAD

**Issue by lencioni** [eslint/eslint#2414]
I'm trying out the new [object-shorthand rule](http://eslint.org/docs/rules/object-shorthand) and have run into a snag in my codebase. We have a number of places where we are using arrow functions as object properties, such as when they are passed in an options object to a function.

For example:

```js
myFunction({
  onComplete: () => this._handleComplete(),
});
```

I would like to be able to turn on this rule but allow arrow functions. I didn't find anything in the documentation for the rule that would enable this.

What do you think?

**Comment by gyandeeps** [comment]
Sounds reasonable to me.
I would like to hear what others think about this.

**Comment by nzakas** [comment]
Can you paste in the output from the console? I'm not quite sure I understand what you're describing.

**Comment by lencioni** [comment]
Sure!

With the following file:

```js
function myFunction(options) {
  if (options.onComplete) {
    options.onComplete();
  }
}

myFunction({
  onComplete: () => this._handleComplete(),
});
```

And the following config:

```yaml
object-shorthand:
  - 2
  - methods
```

ESLint will report the following:

```
myfile.js
  8:2  error  Expected method shorthand

✖ 1 problem (1 error, 0 warnings)
```

**Comment by nzakas** [comment]
Ah, I see. Yeah, we should add an exception for that. Question is if the exception should always be there or if it should be optional. @xjamundx thoughts?

**Comment by xjamundx** [comment]
The only reason I'd be hesitant to make a general exception for arrow functions is because of this mistake, which I've made more than once;

```js
var user = {
   name: "Jamund",
   getName: () => this.name
};
user.getName(); // undefined
```

Alternatives could include opting-out of the rule for that function *or* using `.bind(this)` or avoiding the use of `this` altogether.

I think there's an ES7 proposal that's supposed to address the difficultly of binding that will help address this situation: https://github.com/zenparsing/es-function-bind

So maybe the solution would be to carve out an optional exception for arrow-functions in general or possibly we could do some checking to try to make sure they're doing the right thing like not referencing properties on the same object, but that starts to get really confusing.

Sorry for the lack of a coherent answer, just throwing some thoughts out there!

**Comment by nzakas** [comment]
Sounds like a good plan.

**Comment by xjamundx** [comment]
Great I'm happy to work on it next week...

**Comment by xjamundx** [comment]
Current options: `methods`, `properties`, `never`, `always`

Any suggestions on the new options?

Here are 2 ideas:

1. `[2, "always", "allowArrows"]`
2. `[2, { "apply": "always", "allowArrows": true }]`

We could make #2 backwards compatible with the current `[2, "always"]` style way if we wanted (which I'd suggest we should).

**Comment by bathos** [comment]
The current behavior is seemingly a bug given the description of the rule in the docs; there's no such thing as object literal shorthand for an arrow property. If the scope of "object-shorthand" is expanding to include control over this behavior, then what's proposed above as "allowArrow" might be better in reverse ("disallowArrow") since it's the "augment" to the core rule, not the other way around.

However I'd suggest considering simply making it its own rule. I understand the use case for "disallow arrow in literal," but it doesn't pertain to preferring object shorthand where possible. One is a style concern (object-shorthand), and the other is a language concern ("no-arrow-in-literal"). In the rule index, those two would get categorized differently, so putting them in one rule seems to break with ESlint principles.

**Comment by xjamundx** [comment]
@bathos I think you make a decent argument for keeping them separate. @nzakas is it too late to remove ArrowFunctions from the scope of this rule?

**Comment by nzakas** [comment]
Not too late, we haven't shipped it yet. Now is the time to nail down what we want.

I also find @bathos' reasoning to be very persuasive, so +1

**Comment by xjamundx** [comment]
PR coming to remove the arrow function stuff.
